spec/index.bs

```diff
@@ -665,7 +665,8 @@ enum GPULoadOp {
 };
 
 enum GPUStoreOp {
-    "store"
+    "store",
+    "discard-to-zero"
```

Comment:
I would rather have either "discard" or "zero", but we definitely need this for efficient multisampling.

Reply:
I remember some other options too, "invalidate", "mark-cleared"/"mark-zeroed", and then if we disallowed using uninitialized resources, "discard" or "invalidate"

Reply:
+1 for preferring "discard"
What's the difference between "discard" and "discard-to-zero"?

Reply:
There is no true "discard" as it implies that it is zero-cost and the resource is uninitialized afterward, when in fact it is zeroed (possibly lazily) (like it would be immediately after being created).

Reply:
Seeing this begs the question of ""where's discard".  What's the reason we don't want to allow a true Discard operation? Shouldn't the developers be in control over exactly when the Clear happens (and to what value)?

Reply:
What happens with this is that in the implementations a real discard is performed and the subresource is tagged as needing zero-initialization again. The reason we need to zero init is that the content of the attachment is undefined after a discard and could end up containing data from some other application.

Even if the texture was already cleared, a real discard could allow putting bad data from a previous allocation (and a different web page) because of interaction with texture compression.

The goal is applications could avoid the new lazy-clear by using loadOp clear later which subsumes the clear to zero. If that doesn't happen, the implementation could produce warnings in the console or something.

Reply:
It seems to me that a 'Discard' with the contents being "not anything you shouldn't have access to, but otherwise undefined" would suffice and would provide more flexibility for different implementation choices.

Reply:
This is also my opinion.

Reply:
We have heard this opinion in the group but "undefined but not scary security-wise" is really hard to ensure unless you just make the value defined (see comment about texture compression), also this very issue has been discussed several times and folks agreed to have discard be "as if" clearing to zero, just like resources are created "as if" cleared to zero.

Reply:
Clearing to zero would be a valid implementation choice, but not something we should encourage developers to take a dependency on.  We want developers to fall into the best practice of "discard on end, clear on next load".  Naming it "discard-to-zero" encourages developers to omit the clear on load step as it seems unnecessary.

Reply:
I intend to issue a warning whenever users load from an undefined-data source, so "discard" is the best representation of intent here, to me.

Reply:
(I issue these sorts of warnings already in WebGL today)

Reply:
Another recap of our options:
  1. don't allow uninitialized data (clear resources to zero, have "discard-to-zero")
  2. allow uninitialized data temporarily but require the user to initialize it (hard error if something is discarded and then used without a clear, or just a new resource is used without being initialized)
  3. allow uninitialized data only if it was previously initialized at some point (any garbage is from the user's data, not leaking, not exposing any security issues).

Reply:
I don't think that 3) is an option because when using discard all bets are off. Here's an example that leads to security issues with semi-functional hardware that has the following behavior:

 - There is clear color compression happening, meaning color textures have an additional plane that contains for each texture tile if it is cleared. There is the "regular" plane and the "clear color plane"
 - Render target usage of textures can use the clear color compression, but sampling from the texture can't (or storage usage of the texture).

An Web application creates a texture, does a render pass with `loadOp: "clear"` and `storeOp: "discard"` then samples from the texture in a follow-up pass. What happens in a D3D12 or Vulkan backend is the following:

 - At the beginning of the render pass, the clear color plane is initialized, the regular plane still contains uninitialized data.
 - At the end of the render pass nothing happens
 - Some barriers are emitted but because the texture has been discarded, the barrier is optimized to just wait for the end of the execution of the render pass, and doesn't decompress the clear color plane in the regular plane.
 - The texture is sampled which loads data from the regular plane. Uninitialized data is read, potentially with pixels from your bank account.

This is an example on fictional hardware and you could argue that the barrier optimization is wrong. But the point is that if we rely on undefined behavior in our implementation, some hardware some day will break our assumption that the undefined behavior is safe and will allow malicious web pages to read data it doesn't own.

Reply:
Also we already discussed this issue at length in the group and agreed to go with 1).

Reply:
Good point, @Kangz !

To provide some more context to @damyanp - the problem with (2) was that it would limit the ability of the backend to handle this internally, since the API would have to be conservatively defined. For example, if we copy 2 texel rects into a texture, one for the bottom half and one for the top half, it's tricky to write specification in such a way that this is explicitly allowed.

With this in mind, (1) was considered the only viable option.

Reply:
Option 2 still seems the better option to me.  Is the issue with the conservative spec that you'd like to be able to partially initialize a resource?  I'd be fine with prohibiting this and stating that initialization is either a clear or a full copy.  I'll concede on this though, I've had my say :)

I will add thought that forcing users to read "discard-to-zero" when then mental model should really just be "discard" seems like unnecessary overhead.  A casual developer who is used to using other APIs will need to read the docs to understand what all the "to-zero" stuff is about.  Whereas someone who has thought deeply enough to be concerned about the possible implications of a raw "discard" operation would need to read the docs.  I'd say that the former would outnumber the latter.

Reply:
Keep in mind if we ship the conservative spec today, we can always add a "discard" with looser semantics tomorrow, but if we decide that "discard" was too loose, we won't really add anything with a "discard-to-zero" since we still have to support "discard".

Reply:
My favorite plan for this is that "discard" is meant to yield undefined data, even if initial implementations may clear it.
What I *don't* want, even today, is to entice apps into using "discard-to-zero" to clear things. We should encourage explicit clearing, not reliance on side-effects of today's restricted behavior.

Reply:
We already discussed this as part of issue #97.  The consensus at the time was to call the StoreOp `discard-to-zero`.  

@jdashg , if we want prevent people from forming the bad habit of using `discard-to-zero` to clear things, I think the best we can do is spec things such that the user is required to fill every byte of a discarded texture/buffer with data before they can use it in the pipeline.  Doing this will require more work from implementations that wish to cut corners and eagerly clear discarded (or newly created) resources that the developer fills in multiple steps.